pkg/kubelet/util/pluginwatcher/plugin_watcher.go

```diff
@@ -28,43 +28,137 @@ import (
 	"github.com/pkg/errors"
 	"golang.org/x/net/context"
 	"google.golang.org/grpc"
+
 	registerapi "k8s.io/kubernetes/pkg/kubelet/apis/pluginregistration/v1alpha1"
 	utilfs "k8s.io/kubernetes/pkg/util/filesystem"
 )
 
-// RegisterCallbackFn is the type of the callback function that handlers will provide
-type RegisterCallbackFn func(pluginName string, endpoint string, versions []string, socketPath string) (chan bool, error)
-
 // Watcher is the plugin watcher
 type Watcher struct {
 	path      string
-	handlers  map[string]RegisterCallbackFn
 	stopCh    chan interface{}
 	fs        utilfs.Filesystem
 	fsWatcher *fsnotify.Watcher
 	wg        sync.WaitGroup
-	mutex     sync.Mutex
+
+	mutex       sync.Mutex
+	handlers    map[string]PluginHandler
+	plugins     map[string]pathInfo
+	pluginsPool map[string]map[string]*sync.Mutex // map[pluginType][pluginName]
+}
+
+type pathInfo struct {
+	pluginType string
+	pluginName string
 }
 
 // NewWatcher provides a new watcher
-func NewWatcher(sockDir string) Watcher {
-	return Watcher{
-		path:     sockDir,
-		handlers: make(map[string]RegisterCallbackFn),
-		fs:       &utilfs.DefaultFs{},
+func NewWatcher(sockDir string) *Watcher {
+	return &Watcher{
+		path: sockDir,
+		fs:   &utilfs.DefaultFs{},
+
+		handlers:    make(map[string]PluginHandler),
+		plugins:     make(map[string]pathInfo),
+		pluginsPool: make(map[string]map[string]*sync.Mutex),
 	}
 }
 
-// AddHandler registers a callback to be invoked for a particular type of plugin
-func (w *Watcher) AddHandler(pluginType string, handlerCbkFn RegisterCallbackFn) {
+func (w *Watcher) AddHandler(pluginType string, handler PluginHandler) {
 	w.mutex.Lock()
 	defer w.mutex.Unlock()
-	w.handlers[pluginType] = handlerCbkFn
+
+	w.handlers[pluginType] = handler
 }
 
-// Creates the plugin directory, if it doesn't already exist.
-func (w *Watcher) createPluginDir() error {
+func (w *Watcher) getHandler(pluginType string) (PluginHandler, bool) {
+	w.mutex.Lock()
+	defer w.mutex.Unlock()
+
+	h, ok := w.handlers[pluginType]
+	return h, ok
+}
+
+// Start watches for the creation of plugin sockets at the path
+func (w *Watcher) Start() error {
+	glog.V(2).Infof("Plugin Watcher Start at %s", w.path)
+	w.stopCh = make(chan interface{})
+
+	// Creating the directory to be watched if it doesn't exist yet,
+	// and walks through the directory to discover the existing plugins.
+	if err := w.init(); err != nil {
+		return err
+	}
+
+	fsWatcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to start plugin fsWatcher, err: %v", err)
+	}
+	w.fsWatcher = fsWatcher
+
+	w.wg.Add(1)
```

Comment:
The current construct is:
- The stopChan is used to signal all goroutine that they should stop.
- The Waitgroup is used to wait for all the goroutines to stop

My experience has been that when you are trying to use a single channel to both signal a stop and wait for them to finish, you are basically re-implementing a workgroup usually using a "custom protocol" (sending ints, or constants, ...).

Reply:
In this context, it seems that once the processing loop (inside Start) is signaled to stop (via `<-w.stopCh`), there is no need for further coordination, there's nothing to wait for. Furthermore, any clean up can occur in fswatcher.Close().

Reply:
> In this context, it seems that once the processing loop (inside Start) is signaled to stop (via <-w.stopCh), there is no need for further coordination, there's nothing to wait for. 

How would you catch goroutines that never return or resources that were never freed? This is particularly an issue when you are running tests.

For example if the goroutines that handle plugins of the previous tests are still running (whether you have a bug in your code or by sheer luck), they might interfere with your new test.

It seems like you feel the workgroup is unnecessary. Working in the kubernetes code base I've encountered a lot of issues due to the `Stop` function not freeing all the resources (gRPC being one of them).
On the other hand making sure that a module does indeed completely Stop when we call Stop managed to help catch a lot of bugs and made sure others don't happen.

Reply:
Hi Renaud
I am not arguing against workgroup and there are places, I am saying in this context I don't think you need it because there does not seem to be anything to wait for.  The processing loop in `Start()` has 3 `select` cases

```go
func Start() {...
    for { select {
       // #1
       case event := <-fsWatcher.Events:
           go func()...
      // # 2
       case err := <-fsWatcher.Errors:
           continue
       // # 3
       case <-w.stopCh:
           return
    }}
}
```
`Case 1` runs a new goroutine, and reading through the code, it has no coordination with its caller (no dependency on either channel or workgroup.  Meaning, whether it comes back or die, the caller does not care.
`Case 2` is a simple continue.  There is no coordination needed here, the code does not have to wait for anything.
`Case 3` is our signal watcher.  As soon as `w.stopCh` closes, it wins and ends  the loop and exits immediately.

```go
func Stop() {
    close(w.stopCh)
}
```
So when `Stop()` closes the channel,  `Case 3` will occur and causes the processing loop to return.  At that point, I don't see why now `Stop()` has to wait for anything to complete from the loop above.  Maybe I am missing something...

Reply:
Looks like you caught a bug, I'm not increasing the count when starting a goroutine.

> I am saying in this context I don't think you need it because there does not seem to be anything to wait for. 
> it has no coordination with its caller (no dependency on either channel or workgroup. Meaning, whether it comes back or die, the caller does not care.

You are right in saying that the code in it's current state doesn't need a workgroup. I also agree that there isn't any coordination between the goroutines.

The problem the waitgroup is solving is catching the errors that will arise from changing the code. Making sure that all the gorountines returns protects developers from making mistakes that might not be caught otherwise.
I believe this pattern increases maintainability of the code and provides strong guarantees to the developers using this module (i.e: All resources have been freed and all goroutines are stopped when `Stop()` returns).

Do you think this piece of code decreases the overall readability?

Reply:
> The problem the waitgroup is solving is catching the errors that will arise from changing the code. Making sure that all the gorountines returns protects developers from making mistakes that might not be caught otherwise.

This is a fair point and I get that. It is always good idea to guard against rogue routines with timers to avoid indefinite blocking. 

But using the 1 waitgroup and 2 channels still seems it can be done simply and still achieve the same. For instance, since you are creating a new channel, `c`, in `Stop()` to signal the end of wg.Wait(), how about moving that channel in `Watcher`, let's call it `waitCh` and remove the waitgroup altogether.  So the code becomes: 

```go

type Watcher struct {
    <snip>
    stopCh chan interface{}
    waitCh chan interface{}
}

... 
func (w *Watcher) Start() error {
    <snip>

    w.stopCh = make(chan interface{})  // move in NewWatcher
    w.waitCh = make(chan interface{})  // move in NewWatcher

    ...
    go func(fsWatcher *fsnotify.Watcher) {
        defer close(w.waitCh)

        for { select {
        case event := <-fsWatcher.Events:
           go func()...
        case err := <-fsWatcher.Errors:
           continue
        case <-w.stopCh:
           return
        }}
    }(fsWatcher)
}

...
func (w *Watcher) Stop() error {
	close(w.stopCh)
	
	select {
	case <-w.waitCh:
	case <-time.After(10 * time.Second):
		return fmt.Errorf("timeout on stopping watcher")
	}

	return nil
}
```

Reply:
@vladimirvivien agree that prima facie waitgroup seems redundant but i am in favor of using waitgroup because we saw issues like this one in the past because of lack of proper clean up:
https://github.com/kubernetes/kubernetes/issues/59488
I think using waitgroup is a good pattern to follow, though seems redundant here.

Reply:
> Looks like you caught a bug, I'm not increasing the count when starting a goroutine for the plugins.


+1

Reply:
Sorry for delay.  I will answer this, but in there interest of time, let's keep moving with PR.

> How do you suggest we handle waiting for the multiple goroutines @vladimirvivien?

In the case of work distribution with multiple goroutines, with a need to rendezvous at a certain execution point, a  workgroup certainly makes sense as a synchronization construct.  But the code, as-is in the `Start()` method, is not doing that.  It has 1 goroutine. I was pointing out that it can be simplified, deferring optimizing for multiple goroutines at a later time.